### DISCUSSION TITLE
JSON NESTED CLASSES PROBLEM

### DISCUSSION THREAD

**Discussion by meladam20** [flutterdata/flutter_data#226] [Q&A]
I want to use complex nested classes so i have this two classes

@JsonSerializable(explicitToJson: true)
@DataRepository([JsonServerAdapter])
class User extends DataModel<User> {
  final String? id;
  final String firstName;
  final String lastName;
  final List<Role> role;
}

AND

class Role extends DataModel<Role> {
  final String? id;
  final String libelleRole;
}

All things work well, i can save data in local but  when i use hot reload, i get this error : 

flutter: [flutter_data] Box failed to open:
type '_Map<dynamic, dynamic>' is not a subtype of type 'Map<String, dynamic>'
#0      _$UserFromJson.<anonymous closure> user.g.dart:62

AND 

this is the line 62 of user.g.dart file 

User _$UserFromJson(Map<String, dynamic> json) => User(
      id: json['_id'] as String?,
      firstName: json['firstName'] as String,
      lastName: json['lastName'] as String,
      email: json['email'] as String,
      password: json['password'] as String,
(Line62)role: (json['role'] as List<dynamic>)
          .map((e) => Role.fromJson(e as Map<String, dynamic>))
          .toList(),
    );

Please can you help me ?

**Comment by meladam20** [comment]
Hello @frank06 , please help me. I need an anwser.

**Reply by frank06** [reply]
Use `HasMany`, not `List`

**Reply by meladam20** [reply]
When i use HasMany or either Belong, i can't receive data in my mongo database with my nodejs server. I try it during long time but any issue so i used "List" to progress.

**Reply by frank06** [reply]
Look around https://github.com/flutterdata there is enough sample code, you should be able to make it work

**Reply by meladam20** [reply]
I think that the problem is in my nodejs server

**i change List in HasMany like that :** 

@JsonSerializable(explicitToJson: true)
@DataRepository([JsonServerAdapter])
class User extends DataModel {
final String? id;
final String firstName;
final String lastName;
final HasMany<Role> role;

User({
    this.id,
    required this.firstName,
    required this.lastName,
    required this.email,
    required this.password,
    required this.role,
  });
}

**And the user that i want to save in mongo database :**

final developerRole = Role(libelleRole: 'developpeur');

    final user = User(
        firstName: firstName.text,
        lastName: lastName.text,
        email: email.text,
        password: password.text,
        role: HasMany({developerRole}));

user.save();


**Look at my mongoose model**

const UserSchema = new mongoose.Schema({
   firstName: {
        type: String,
        required: true
    },
    lastName: {
        type: String,
        required: true
    },
    email: {
        type: String,
        required: true
    },
    password: {
        type: String,
        required: true
    },
    role: [{
        type: 'ObjectId',
        ref: 'Role',
    }]
},
{versionKey: false}
);

**But in my user.controller.js of the nodejs, i can't get the value of the field "Role" in req.body**

const addUser = async (req, res) => {
    console.log(req.body);
const newUser = new User({...req.body});
let userSave = await newUser.save();
res.json(userSave);
};

**In the console, the value of req.body is :** 

{
  role: [],
  _id: 648caa6429ad10965a0c455b,
  firstName: 'zer',
  lastName: 'zef',
  email: 'aqsz@gmail.com',
  password: '$2a$10$QOIiUv3ExqhB6RdjjBWCj.JGNUeayFraXdjTdkgluPQTCqFDTkM42'
}

**Please why the value of array role is always empty despite all my tests  ?**

**Reply by meladam20** [reply]
Please @frank06, this is a right way to save model witth hasMany relationship ?

final developerRole = Role(libelleRole: 'developpeur');

final user = User(
    firstName: firstName.text,
    lastName: lastName.text,
    email: email.text,
    password: password.text,
    role: HasMany({developerRole}));
user.save();

I just want to be sure that is right way to remote this kind of data.

**Reply by frank06** [reply]
I don't have time to go through you code, I'm sorry

If you want to save relationship data you definitely need to override the serializers. Please search the issues https://github.com/flutterdata/flutter_data/issues , I'm sure I explained that a few times. I might have explained it to you in one of yours.

Here's a good example of custom serializers (just for the structure): https://github.com/flutterdata/flutter_data_json_api_adapter

If not just set up your debugger and step through FD code, it's not hard to understand the whole process
